File: sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/internal/TaskNodeOutputImpl.java
package io.littlehorse.sdk.wfsdk.internal;

import io.littlehorse.sdk.common.proto.ExponentialBackoffRetryPolicy;
import io.littlehorse.sdk.wfsdk.TaskNodeOutput;

public class TaskNodeOutputImpl extends NodeOutputImpl implements TaskNodeOutput {

    public TaskNodeOutputImpl(String nodeName, WorkflowThreadImpl parent) {
        super(nodeName, parent);
    }

    @Override
    public TaskNodeOutputImpl withExponentialBackoff(ExponentialBackoffRetryPolicy policy) {
        parent.overrideTaskExponentialBackoffPolicy(this, policy);
        return this;
    }

    @Override
    public TaskNodeOutputImpl withRetries(int retries) {
        parent.overrideTaskRetries(this, retries);
        return this;
    }

    @Override
    public TaskNodeOutput timeout(int timeoutSeconds) {
        parent.addTimeoutToTaskNode(this, timeoutSeconds);
        return this;
    }
}

File: sdk-js/src/workflow/internal/taskNodeOutputImpl.ts
import type { ExponentialBackoffRetryPolicy } from '../../proto/common_wfspec'
import type { TaskNodeOutput } from '../taskNodeOutput'
import { NodeOutputImpl } from './nodeOutputImpl'
import type { WorkflowThreadImpl } from './workflowThreadImpl'

export class TaskNodeOutputImpl extends NodeOutputImpl implements TaskNodeOutput {
  constructor(nodeName: string, thread: WorkflowThreadImpl) {
    super(nodeName, thread)
  }

  withRetries(retries: number): TaskNodeOutput {
    this.thread.patchTaskNode(this.nodeName, t => {
      t.retries = retries
    })
    return this
  }

  withTimeout(timeoutSeconds: number): TaskNodeOutput {
    this.thread.patchTaskNode(this.nodeName, t => {
      t.timeoutSeconds = timeoutSeconds
    })
    return this
  }

  withExponentialBackoff(policy: ExponentialBackoffRetryPolicy): TaskNodeOutput {
    this.thread.patchTaskNode(this.nodeName, t => {
      t.exponentialBackoff = policy
    })
    return this
  }
}
